File: 프로그래머스/4/118670. 행렬과 연산/행렬과 연산.cpp
#include <string>
#include <vector>
#include <deque>
#include <iostream>
#include <utility>

using namespace std;

vector<vector<int>> solution(vector<vector<int>> rc, vector<string> operations) {
    int r = rc.size();
    int c = rc[0].size();
    
    vector<vector<int>> answer(r, vector<int>(c,0));
    deque<int> left, right;
    deque<deque<int>> mid;
    
    for(int i=0;i<r;i++) {
        left.push_back(rc[i][0]);
        right.push_back(rc[i][c-1]);
        
        deque<int> tmp;
        for(int j=1;j<c-1;j++) {
            tmp.push_back(rc[i][j]);
        }
        mid.push_back(tmp);
    }
    
    for(auto op: operations) {
        if (op == "ShiftRow") {
            left.push_front(left.back());
            left.pop_back();
            mid.push_front(std::move(mid.back()));
            mid.pop_back();
            right.push_front(right.back());
            right.pop_back();
        } else { // Rotate
            if(c > 2) {
                mid.front().push_front(left.front());
                left.pop_front();

                right.push_front(mid.front().back());
                mid.front().pop_back();

                mid.back().push_back(right.back());
                right.pop_back();

                left.push_back(mid.back().front());
                mid.back().pop_front();                
            } else {
                right.push_front(left.front());
                left.pop_front();
                
                left.push_back(right.back());
                right.pop_back();
            }
        }
    }
    
    for(int i=0;i<r;i++) {
        answer[i][0] = left[i];
        answer[i][c-1] = right[i];
        
        for(int j=1;j<c-1;j++) {
            answer[i][j] = mid[i][j-1];
        }
    }
    return answer;
}

File: 프로그래머스/4/118670. 행렬과 연산/행렬과 연산.py
from collections import deque

def solution(rc, operations):
    answer = []
    
    left = deque()
    right = deque()
    mid = deque()
    
    n = len(rc)
    m = len(rc[0])
    
    for i in range(n):
        left.append(rc[i][0])
        right.append(rc[i][m-1])
        
        tmp = deque()
        for j in range(1, m-1):
            tmp.append(rc[i][j])
        mid.append(tmp)
    
    for op in operations:
        if op == "ShiftRow":
            left.appendleft(left.pop())
            right.appendleft(right.pop())
            mid.appendleft(mid.pop())
        else:
            mid[0].appendleft(left.popleft())
            right.appendleft(mid[0].pop())
            mid[n-1].append(right.pop())
            left.append(mid[n-1].popleft())
    
    for i in range(n):
        tmp = []
        tmp.append(left.popleft())
        while mid[i]: tmp.append(mid[i].popleft())
        tmp.append(right.popleft())
        answer.append(tmp)
    
    return answer